src/libopensc/card-idprime.c

```diff
@@ -65,33 +85,84 @@ static const sc_path_t idprime_path = {
 typedef struct idprime_object {
 	int fd;
 	unsigned char key_reference;
+	int valid_key_ref;
 	u8 df[2];
 	unsigned short length;
+	int pin_index;
 } idprime_object_t;
 
+/*
+ * IDPrime Container structure
+ * Simplification of auxiliary data from aux-data.c
+ */
+#define MAX_CONTAINER_NAME_LEN 39
+#define CONTAINER_OBJ_LEN 86
+typedef struct idprime_container {
+	int index;								/* Index of the container */
+	char guid[MAX_CONTAINER_NAME_LEN + 1];	/* Container name */
+	struct idprime_container *next;
+} idprime_container_t;
+
+/*
+ * IDPrime key reference structure
+ */
+#define KEYREF_OBJ_LEN 8
+typedef struct idprime_keyref {
+	int index;								/* Index of the key reference */
+	int pin_index;							/* Index of the auth pin used for accessing key */
+	unsigned char key_reference;			/* Key reference used for accessing key */
+	struct idprime_keyref *next;
+} idprime_keyref_t;
+
 /*
  * IDPrime private data per card state
  */
 typedef struct idprime_private_data {
-	u8 *cache_buf;			/* cached version of the currently selected file */
-	size_t cache_buf_len;		/* length of the cached selected file */
-	int cached;			/* is the cached selected file valid */
-	size_t file_size;		/* this is real file size since IDPrime is quite strict about lengths */
-	list_t pki_list;		/* list of pki containers */
-	idprime_object_t *pki_current;	/* current pki object _ctl function */
-	int tinfo_present;		/* Token Info Label object is present*/
-	u8 tinfo_df[2];			/* DF of object with Token Info Label */
+	u8 *cache_buf;				/* cached version of the currently selected file */
+	size_t cache_buf_len;			/* length of the cached selected file */
+	int cached;				/* is the cached selected file valid */
+	size_t file_size;			/* this is real file size since IDPrime is quite strict about lengths */
+	list_t pki_list;			/* list of pki containers */
+	idprime_object_t *pki_current;		/* current pki object _ctl function */
+	int tinfo_present;			/* Token Info Label object is present*/
+	u8 tinfo_df[2];				/* DF of object with Token Info Label */
+	unsigned long current_op;		/* current operation set by idprime_set_security_env */
+	idprime_container_t *containers;	/* list of private key containers */
+	idprime_keyref_t *keyrefmap;		/* list of key references for private keys */
 } idprime_private_data_t;
 
 /* For SimCList autocopy, we need to know the size of the data elements */
 static size_t idprime_list_meter(const void *el) {
 	return sizeof(idprime_object_t);
 }
 
+static void idprime_free_containermap(idprime_container_t *containers)
+{
+	idprime_container_t *next = NULL;
+	while (containers) {
+		next = containers->next;
+		free(containers);
+		containers = next;
+	}
+}
+
+
+static void idprime_free_keyref(idprime_keyref_t *keyrefmap)
+{
+	idprime_keyref_t *next = NULL;
+	while (keyrefmap) {
+		next = keyrefmap->next;
+		free(keyrefmap);
+		keyrefmap = next;
+	}
+}
```

Comment:
idprime_keyref_t and idprime_container_t are defining list types. Wouldn't it be better to use the existing list types from simclist?

note that there is also https://github.com/OpenSC/OpenSC/issues/2127

Reply:
Yes, rewritten in 6eb67f876c203c615498016c255036f9e1dd15d5.